### DISCUSSION TITLE
Duplicate symbol definition linker error

### DISCUSSION THREAD

**Discussion by simone-gaiarin** [nlohmann/json#2493] [Q&A] [answered]
When I include `<nlohmann/json.hpp>` in two different header files I get the multiple linker errors regarding duplicate symbol definition (I just report one of them):

```
[build] A.cpp.obj : warning LNK4006: "void __cdecl nlohmann::detail::dtoa_impl::grisu2_round(char *,int,unsigned __int64,unsigned __int64,unsigned __int64,unsigned __int64)" (?grisu2_round@dtoa_impl@detail@nlohmann@@YAXPADH_K111@Z) already defined in B.cpp.obj; second definition ignored
```

```
[build] lld-link: error: duplicate symbol: void __cdecl nlohmann::detail::dtoa_impl::grisu2_round(char *, int, unsigned __int64, unsigned __int64, unsigned __int64, unsigned __int64)
```

I defined a file called `"json.hpp"` with the following content:
```
#pragma once

#include <nlohmann/json.hpp>

using json = nlohmann::json;
```

then I import it as follow:

A.hpp
```
#pragma once

#include "json.hpp"
```

B.hpp
```
#pragma once

#include "json.hpp"
#include "A.hpp"
```

Relevant part of `CMakeLists.txt`
```
set(JSON_MultipleHeaders ON)

include(FetchContent)

FetchContent_Declare(json
  GIT_REPOSITORY https://github.com/nlohmann/json
  GIT_TAG v3.7.3
)

FetchContent_GetProperties(json)
if(NOT json_POPULATED)
  FetchContent_Populate(json)
  add_subdirectory(${json_SOURCE_DIR} ${json_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()

add_library(libsipnativemessaginghost STATIC 
	"A.cpp"
	"B.cpp"
)
target_link_libraries(myproject PUBLIC nlohmann_json::nlohmann_json)
```

Any idea what can cause the issue or what can I do to debug it further?

I tried to use `json_fwd.hpp` but given that I have a lines like this in both header, importing `json_fwd.hpp` is not enough.

```
void method(const json &param1= json::value_t::object) const;
```

I tried to replicate this isssue in a dummy project, but there everything works as expected, even if I include `<nlohmann/json.hpp>` multiple times in the headers.

**Comment by nlohmann** [comment, marked as answer]
I am working in the exact setup at work and don’t have such an error. 🤔

**Reply by simone-gaiarin** [reply]
Is it there a non-header only version of the library that I can use to avoid including it in the headers?

**Reply by nlohmann** [reply]
No, it’s header only. But I never heard of such problems.

**Reply by simone-gaiarin** [reply]
Found the source of the problem in an third party library
```
/* Define 'inline' to nothing, since the MSVC compiler doesn't support it.  */
#define inline
```

It was completely disrupting the header-only library mechanism.

**Reply by nlohmann** [reply]
Wow, that’s subtle! Thanks for sharing!

**Reply by gregmarr** [reply]
I was debating whether or not I was crazy to ask if there was something doing something insane like defining away `inline`, since that's the only way I could see to get that error.  I guess I wasn't so crazy after all.
